### DISCUSSION TITLE
Battery Voltage detection upto 2 decimals

### DISCUSSION THREAD

**Discussion by kakashi2018** [ExpressLRS/ExpressLRS#2063] [General]
Is it possible to make the cell voltage detection to 2 decimals  for elrs? I know that CRSF spec is only for 1 decimal but when using a 1s lipo for micro it is really usefull to get the 2 decimal precision on lipos

**Comment by Combinacijus** [comment]
I totally agree that we need 2 decimal places for voltage and not only on 1S, on any S because it's per cell basis. And difference between 3.6V and 3.5V per cell is massive it's tens of % of battery capacity and 3.4V already damages lipos. So there is no good point of setting low voltage alarm on a radio. Either you leave tens of percent of battery not used or you over-discharged it.

This comment add some info. https://github.com/ExpressLRS/ExpressLRS/discussions/1464#discussioncomment-2277553

> The problem with report_cell_voltage is that CRSF protocol sucks donkey spit because there's only one decimal place for VBAT precision (but supports up to 6553.5V because of course TBS always has the best ideas). If you want two digits of precision you're going to have to take that request up the chain to TBS and everyone else everywhere that displays a VBAT value (or use a lower ANALOG_VBAT_SCALE).

Betaflight implementation crsf.c might be helpful: https://github.com/betaflight/betaflight/blob/daa6df80248c9a806b7d77c402d415a15f4e2667/src/main/telemetry/crsf.c#L191-L216

There's definitely no need for 6553.5V max voltage, better make it 655.35V
- 655.35V is plenty enough with 0.01V resolution
- meanwhile 65.535V is around 15S which is not that crazy to think that someone would exceed that in the future and we might not need 0.001V resolution

Does anyone know how hard realistically it would be to change it? What if ELRS didn't use CRSF and made it's own protocol or fork of it

Is it really that hopeless quote above says?

**Comment by offer-shmuely** [comment]
I like that too
Following…

**Comment by arhifisyant** [comment]
vote up for this. Just randomly browsing for this idea, and found this discussion. Come on, this is a must. its about precision. 3.6 V and 3.5 V is a huge difference

**Reply by offer-shmuely** [reply]
It will help for voltage calibration as well

**Comment by silicongarage** [comment]
Been down this path earlier and vote for adding more precision. 
 
As a workaround/hack I've been upscaling the VBat voltage 10x before the ELRS receiver and then down scaling by 10x on TX16S.   Been using this for a while on a couple different powered gliders and it works well, but as always YMMV.

Pretty straight forward on iNav

![vbat-inav](https://github.com/ExpressLRS/ExpressLRS/assets/35607907/c908bb9f-9946-4163-8a85-682caac8e92f)

For non-FC,  connect your ELRS receiver to WiFi then go to "elrs_rx.local/hardware.html" and you will find settings to scale VBat.
I haven't tried this yet, as I have no PWM receivers

![vbat-elrs_rx local:hardware html](https://github.com/ExpressLRS/ExpressLRS/assets/35607907/a3cfba39-4305-4719-9ca2-e74317886565)

Because of the way EdgeTX handles sensor scaling, you really can't get an exact divide by ten.  I just created a side branch of the EdgeTX src tree and patched "crossfire.cpp" to handle the 'effective' precision

![crossfire cpp](https://github.com/ExpressLRS/ExpressLRS/assets/35607907/8de408ed-f237-4374-823b-1a6ea1357a24)

Then for the sensor set the precision to 2 places

![screen-2024-07-01-133539](https://github.com/ExpressLRS/ExpressLRS/assets/35607907/469f957c-1ec8-49cc-b981-94a0ff18dee1)

**Reply by arhifisyant** [reply]
Thanks. Youre inspiring me to do the same workaround. Ive been thinking about this also, but not sure. Because it will mess with the OSD  vbat. On the OSD it will appear hundreds volt. Its sketchy, we get our precision voltage monitor on the transmitter, but we mess the OSD. 

Update : 
by the way, in the main branch, installed on my zorro, we already have precision 0.00 in the edgetx itself. The default is 0.0 but we can pick 0.00 manually.

**Reply by silicongarage** [reply]
I fly only LOS but can understand the problem with upscaling VBat in the FC then displaying in OSD.

If I remember correctly when I first worked on this issue the EdgeTX precision field (0.00) only seemed to change the VBat output format, and  just set the second digit to zero.  I had to patch crossfire.cpp to indicate that the incoming telemetry stream was effectively 2 digits.

**Reply by arhifisyant** [reply]
> I fly only LOS but can understand the problem with upscaling VBat in the FC then displaying in OSD.
> 
> If I remember correctly when I first worked on this issue the EdgeTX precision field (0.00) only seemed to change the VBat output format, and just set the second digit to zero. I had to patch crossfire.cpp to indicate that the incoming telemetry stream was effectively 2 digits.

Ahaa. I see now. Didnt know that yet. Thank you for the info.

**Reply by edemargerie** [reply]
FWIW, here are the default values I have on an ER6GV : 
VBat pin = 39
VBat offset = -62
VBat scale = 1024   _(comment : voltage = (analog - offset) / scale)_
VBat attenuation = 11 dB + calibration.

should I change scale to 102 to use the present workaround/hack ?

**Reply by silicongarage** [reply]
@edemargerie  I setup an ER6 a while back for another person,  and if I remember correctly I did just that, lower the Vbat scale to 1/10 of the original.   I do remember that the end result didn't seem that accurate but I didn't spend a lot of time on the problem.   I did simplify the TX backend as I got tired of constantly patching every new EdgeTX release.   Wrote a Lua function that reads Vbat, divides by 10 and creates a  new virtual telemetry sensor with the result.

**Comment by druckgott** [comment]
I would prefer this in the normal telemetry to get minimum 1, better 2 decimales and do not have any workaround for this. Voltage for a Lipo is very important to not destry it.

**Reply by silicongarage** [reply]
The single digit precision is defined as part of the CRSF protocol  
> https://github.com/crsf-wg/crsf/wiki/Packet-Types

As CRSF was initial developed for drones with high current/high voltage it seems precision was deemed not that important.  With fixed wing gliders I use much lower voltage and the motor is off most of the time, so the extra precision on the battery voltage is quite useful.   From what I understand, at some point there was discussion about creating an extended ELRS CRSF telemetry protocol to allow for more precision and additional sensors.  Not sure what the current status is. Hopefully with ELRS finding more application in fixed wings there will be more interest in fixing this problem.

**Comment by edemargerie** [comment]
Digging this up, has 2 decimal precision been implemented in newer ELRS versions (4.x ?)
There are now sensors that can report 0.01V https://radiomasterrc.com/collections/ers-series-sensor/products/ers-cv01-precision-cell-voltage-sensor-for-compatible-expresslrs-pwm-receivers
